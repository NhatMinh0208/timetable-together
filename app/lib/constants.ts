export const TICKS_IN_DAY = 24 * 60 * 60 * 1000;
export const TICKS_IN_WEEK = 7 * TICKS_IN_DAY;

export const weekdays = ["MON", "TUE", "WED", "THU", "FRI", "SAT", "SUN"];

export const STATUS_PENDING = "pending";
export const STATUS_ACTIVE = "active";

export const EVENT_PAGE_SIZE = 15;
