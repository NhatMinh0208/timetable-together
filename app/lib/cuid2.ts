import { init } from "@paralleldrive/cuid2";

// The init function returns a custom createId function with the specified
// configuration. All configuration properties are optional.
export const createId10 = init({
  // A custom random function with the same API as Math.random.
  // You can use this to pass a cryptographically secure random function.
  random: Math.random,
  // the length of the id
  length: 10,
  // A custom fingerprint for the host environment. This is used to help
  // prevent collisions when generating ids in a distributed system.
  fingerprint: "a-custom-host-fingerprint",
});
