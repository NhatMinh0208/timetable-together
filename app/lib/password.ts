import * as bcrypt from "bcryptjs";
const saltRounds = 10;

export async function saltAndHashPassword(password: string) {
  return await bcrypt.hash(password, saltRounds);
}

export async function passwordMatch(password: string, pwHash: string) {
  return await bcrypt.compare(password, pwHash);
}
