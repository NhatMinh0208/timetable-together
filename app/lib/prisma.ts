import { PrismaClient } from "@prisma/client";
// import { withPulse } from "@prisma/extension-pulse";

const globalForPrisma = globalThis as unknown as { prisma: PrismaClient };

export const prisma = globalForPrisma.prisma || new PrismaClient();
// .$extends(
//   withPulse({
//     apiKey: process.env["PULSE_API_KEY"] as string,
//   }),
// );

if (process.env.NODE_ENV !== "production") globalForPrisma.prisma = prisma;
