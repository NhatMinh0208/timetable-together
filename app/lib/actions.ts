"use server";

import { passwordMatch, saltAndHashPassword } from "@/app/lib/password";
import {
  insertUserToDb,
  getUserFromDb,
  getEventFromName,
  getAttendance,
  insertAttendance,
  getAttendancesByUserId,
  updateAttedance,
  removeAttendance,
} from "@/app/lib/db";
import { ExtendedAttendance, User } from "@/app/lib/types";
import { auth } from "@/auth";
import { z } from "zod";
import { revalidatePath } from "next/cache";
export async function createUser(
  email: string,
  name: string,
  password: string,
) {
  const pwHash = await saltAndHashPassword(password);
  await insertUserToDb(email, name, pwHash);
}

export async function authUser(
  email: string,
  password: string,
): Promise<User | null> {
  const user = await getUserFromDb(email);
  if (user == null) throw new Error("User not found");
  if (await passwordMatch(password, user.password)) {
    return {
      id: user.id,
      email: user.email,
      name: user.name,
    };
  } else {
    throw new Error("Password mismatch");
  }
}

export async function getUser(email: string): Promise<User | null> {
  const user = await getUserFromDb(email);
  if (user == null) throw new Error("User not found");
  return {
    id: user.id,
    email: user.email,
    name: user.name,
  };
}

const AddAttendanceSchema = z.object({
  eventName: z.string(),
});

export type AddAttendanceState =
  | {
      errors?: {
        eventName?: string[];
      };
      message?: string;
    }
  | undefined;

export async function addAttendance(
  prevState: AddAttendanceState,
  formData: FormData,
) {
  try {
    const session = await auth();
    if (!session)
      return {
        message: "Failed to add event to timetable: User is not logged in",
      };
    const userId = session?.user?.id ? session.user.id : "";
    const eventName = formData.get("eventName");
    const fields = AddAttendanceSchema.safeParse({
      eventName: formData.get("eventName"),
    });
    if (!fields.success) {
      return {
        errors: fields.error.flatten().fieldErrors,
        message: "Failed to add event to timetable: Form Error",
      };
    }
    const event = await getEventFromName(fields.data.eventName);
    if (!event)
      return {
        message: "Failed to add event to timetable: No such event exists",
      };
    if (event.schedules.length === 0)
      return {
        message: "Failed to add event to timetable: Event has no schedules",
      };
    const attendance = await getAttendance(userId, event.id);
    if (attendance)
      return {
        message:
          "Failed to add event to timetable: Event is already in timetable",
      };
    await insertAttendance(userId, event.id, event.schedules[0].id);
    revalidatePath("/timetable");
  } catch (error) {
    console.log(error);
    return {
      message: "Failed to add event to timetable: Database Error",
    };
  }
}

export async function getUserAttendances(): Promise<ExtendedAttendance[]> {
  try {
    const session = await auth();
    if (!session) {
      throw new Error("User is not logged in");
    }
    const userId = session?.user?.id ? session.user.id : "";
    return await getAttendancesByUserId(userId);
  } catch (error) {
    console.log(error);
    return [];
  }
}

export async function updateUserAttendance(
  eventId: string,
  scheduleId: string,
): Promise<void> {
  try {
    const session = await auth();
    if (!session) {
      throw new Error("User is not logged in");
    }
    const userId = session?.user?.id ? session.user.id : "";
    await updateAttedance(userId, eventId, scheduleId);
    console.log("done");
  } catch (error) {
    console.log(error);
    return;
  }
}

export async function removeUserAttendance(eventId: string): Promise<void> {
  try {
    const session = await auth();
    if (!session) {
      throw new Error("User is not logged in");
    }
    const userId = session?.user?.id ? session.user.id : "";
    await removeAttendance(userId, eventId);
    console.log("done");
    revalidatePath("/timetable");
  } catch (error) {
    console.log(error);
    return;
  }
}
