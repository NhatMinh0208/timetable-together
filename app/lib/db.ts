"use server";

import { PrismaClient, Prisma } from "@prisma/client";
import { createId10 } from "@/app/lib/cuid2";
import { ExtendedAttendance } from "@/app/lib/types";
const prisma = new PrismaClient();

export async function getUserFromDb(email: string) {
  return await prisma.user.findUnique({
    where: {
      email: email,
    },
  });
}

export async function insertUserToDb(
  email: string,
  name: string,
  password: string,
) {
  try {
    await prisma.user.create({
      data: {
        id: createId10(),
        email: email,
        name: name,
        password: password,
      },
    });
  } catch (error) {
    console.log("Database error!");
    console.log(error);
  }
}

export async function getEventFromName(name: string) {
  return await prisma.event.findFirst({
    where: {
      name: name,
    },
    select: {
      id: true,
      name: true,
      schedules: {
        select: {
          id: true,
        },
      },
    },
  });
}

export async function getAttendance(userId: string, eventId: string) {
  return await prisma.attendance.findUnique({
    where: {
      attendeeId_eventId: {
        attendeeId: userId,
        eventId: eventId,
      },
    },
  });
}

export async function insertAttendance(
  userId: string,
  eventId: string,
  scheduleId: string,
) {
  return await prisma.attendance.create({
    data: {
      attendeeId: userId,
      eventId: eventId,
      scheduleId: scheduleId,
    },
  });
}

export async function updateAttedance(
  userId: string,
  eventId: string,
  scheduleId: string,
) {
  return await prisma.attendance.update({
    where: {
      attendeeId_eventId: {
        attendeeId: userId,
        eventId: eventId,
      },
    },
    data: {
      scheduleId: scheduleId,
    },
  });
}

export async function removeAttendance(userId: string, eventId: string) {
  return await prisma.attendance.delete({
    where: {
      attendeeId_eventId: {
        attendeeId: userId,
        eventId: eventId,
      },
    },
  });
}

export async function getAttendancesByUserId(
  userId: string,
): Promise<ExtendedAttendance[]> {
  return await prisma.attendance.findMany({
    where: {
      attendeeId: userId,
    },
    select: {
      event: {
        select: {
          id: true,
          name: true,
          schedules: {
            select: {
              id: true,
              name: true,
              sessions: true,
            },
          },
        },
      },
      scheduleId: true,
      attendeeId: true,
    },
    orderBy: {
      event: {
        name: "asc",
      },
    },
  });
}
