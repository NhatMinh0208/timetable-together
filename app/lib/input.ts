import {
  EventInput,
  ExtendedEvent,
  ExtendedSchedule,
  ScheduleInput,
  Session,
  SessionInput,
} from "@/app/lib/types";
import { eventSchema, scheduleSchema, sessionSchema } from "@/app/lib/zod";
import { CreateEventState } from "@/app/lib/actions";
import { ZodError } from "zod";
import { errorToJSON } from "next/dist/server/render";

export function convertSessionInput(
  state: CreateEventState,
  input: SessionInput,
  schedule: string,
  session: number,
): Session {
  try {
    const parseResult = sessionSchema.parse(input);
  } catch (e) {
    if (e instanceof ZodError) {
      for (const err of e.errors) {
        state.errors.push(
          "Schedule " +
            schedule +
            ", session #" +
            session.toString() +
            ": " +
            err.message,
        );
      }
    }
  }
  const [startDay, startMonth, startYear] = input.startDate
    .split("/")
    .map((s) => parseInt(s, 10));
  const [endDay, endMonth, endYear] = input.endDate
    .split("/")
    .map((s) => parseInt(s, 10));
  return {
    id: "",
    place: input.place,
    timeZone: input.timeZone,
    startDate: new Date(startYear, startMonth - 1, startDay, 0, 0, 0, 0),
    endDate: new Date(endYear, endMonth - 1, endDay, 0, 0, 0, 0),
    startTime: new Date("01/01/2024 " + input.startTime),
    endTime: new Date("01/01/2024 " + input.endTime),
    interval: parseInt(input.interval),
  };
}

function convertScheduleInput(
  state: CreateEventState,
  input: ScheduleInput,
): ExtendedSchedule {
  try {
    const parseResult = scheduleSchema.parse(input);
  } catch (e) {
    if (e instanceof ZodError) {
      for (const err of e.errors) {
        state.errors.push("Schedule " + input.name + ": " + err.message);
      }
    }
  }
  return {
    id: "",
    name: input.name,
    sessions: input.sessions.map((x, i) =>
      convertSessionInput(state, x, input.name, i + 1),
    ),
  };
}

export function convertEventInput(
  state: CreateEventState,
  input: EventInput,
): ExtendedEvent {
  try {
    const parseResult = eventSchema.parse(input);
  } catch (e) {
    if (e instanceof ZodError) {
      for (const err of e.errors) {
        state.errors.push(err.message);
      }
    }
  }
  return {
    id: "",
    name: input.name,
    description: input.description,
    private: input.private,
    schedules: input.schedules.map((x) => convertScheduleInput(state, x)),
  };
}

function pad(s: string, digits: number) {
  let res = s;
  while (res.length < digits) {
    res = "0" + res;
  }
  return res;
}

function convertSessionToInput(session: Session): SessionInput {
  return {
    id: session.id,
    place: session.place,
    timeZone: session.timeZone,
    startDate:
      pad(session.startDate.getDate().toString(), 2) +
      "/" +
      pad((session.startDate.getMonth() + 1).toString(), 2) +
      "/" +
      session.startDate.getFullYear().toString(),
    endDate:
      pad(session.endDate.getDate().toString(), 2) +
      "/" +
      pad((session.endDate.getMonth() + 1).toString(), 2) +
      "/" +
      session.endDate.getFullYear().toString(),
    startTime:
      pad(session.startTime.getHours().toString(), 2) +
      ":" +
      pad(session.startTime.getMinutes().toString(), 2),
    endTime:
      pad(session.endTime.getHours().toString(), 2) +
      ":" +
      pad(session.endTime.getMinutes().toString(), 2),
    interval: session.interval.toString(),
  };
}

function convertScheduleToInput(schedule: ExtendedSchedule): ScheduleInput {
  return {
    id: schedule.id,
    name: schedule.name,
    sessions: schedule.sessions.map(convertSessionToInput),
  };
}

export function convertEventToInput(event: ExtendedEvent): EventInput {
  return {
    name: event.name,
    description: event.description,
    private: event.private,
    schedules: event.schedules.map(convertScheduleToInput),
  };
}
