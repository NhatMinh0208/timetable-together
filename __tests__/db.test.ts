import "@testing-library/jest-dom";
import * as db from "@/app/lib/db";
import { Event, Schedule, User } from "@prisma/client";
import { prisma } from "@/app/lib/prisma";
import {
  ExtendedAttendance,
  ExtendedEvent,
  ExtendedSchedule,
} from "@/app/lib/types";

// helper functions

function purgeId(
  event:
    | ExtendedEvent
    | (ExtendedEvent & {
        owner: {
          id: string;
          name: string;
        };
      })
    | {
        id: string;
        name: string;
        owner: {
          name: string;
        };
      },
) {
  event.id = "";
  if ("owner" in event && "id" in event.owner) {
    event.owner.id = "";
  }
  if ("schedules" in event) {
    for (const schedule of event.schedules) {
      schedule.id = "";
      for (const session of schedule.sessions) {
        session.id = "";
        (session as unknown as { scheduleId: string }).scheduleId = "";
      }
    }
  }
}

function purgeIdAttendance(attendance: ExtendedAttendance) {
  attendance.attendeeId = "";
  purgeId(attendance.event);
  attendance.scheduleId = "";
}

export async function createEventStandalone(
  event: ExtendedEvent,
  userId: string,
) {
  try {
    const insertedEvent = await db.insertEvent(
      userId,
      event.name,
      event.description,
      event.private,
    );
    for (const schedule of event.schedules) {
      const insertedSchedule = await db.insertSchedule(
        insertedEvent.id,
        schedule.name,
      );
      for (const session of schedule.sessions) {
        await db.insertSession(
          insertedSchedule.id,
          session.place,
          session.timeZone,
          session.startTime,
          session.endTime,
          session.startDate,
          session.endDate,
          session.interval,
        );
      }
    }
    return insertedEvent;
  } catch (error) {
    console.log(error);
  }
}

// test data

const testUsers: User[] = [
  {
    id: "",
    name: "user_f8f8fsdf0",
    email: "user22@test.example.com",
    password: "password",
  },
  {
    id: "",
    name: "用户_bcvnft908nf",
    email: "yonghu22@test.example.org",
    password: "mima1234",
  },
  {
    id: "",
    name: "người dùng_hjmghj09m",
    email: "nguoidung22@test.example.org",
    password: "matkhau1",
  },
];

const testEventsOwner = testUsers[0];

const testSchedule: ExtendedSchedule = {
  id: "",
  name: "Test Schedule #1",
  sessions: [
    {
      id: "",
      place: "Test Location",
      timeZone: "",
      interval: 7,
      startDate: new Date(1720803600000),
      endDate: new Date(1731430800000),
      startTime: new Date(1704070800000),
      endTime: new Date(1704103200000),
    },
  ],
};
const testScheduleB: ExtendedSchedule = {
  id: "",
  name: "Test Schedule #2",
  sessions: [
    {
      id: "",
      place: "Test Location B",
      timeZone: "",
      interval: 7,
      startDate: new Date(1720803600000),
      endDate: new Date(1731430800000),
      startTime: new Date(1704070800000),
      endTime: new Date(1704103200000),
    },
  ],
};

const testEvents: ExtendedEvent[] = [
  {
    id: "",
    name: "CS9999 Introduction to Testing - Lecture (Sem 1)",
    description: "A test case for fetching an event with an exact name",
    private: false,
    schedules: [testSchedule],
  },
  {
    id: "",
    name: "CS9999 Introduction to Testing - Tutorial (Sem 1)",
    description: "A test case for fetching multiple events with a prefix",
    private: true,
    schedules: [testSchedule],
  },
  {
    id: "",
    name: "CS9999 Introduction to Testing - Tutorial (Sem 2)",
    description: "A test case for fetching multiple events with a prefix",
    private: true,
    schedules: [testSchedule],
  },
  {
    id: "",
    name: "CS1244 Multi-Schedule test",
    description: "A test case for updating attendance",
    private: false,
    schedules: [testSchedule, testScheduleB],
  },
];

// initialization and cleanup for this suite

beforeAll(async () => {
  for (const test of testEvents) {
    const events = await db.getEventsFromName(
      test.name,
      20,
      true,
      testUsers[0].id,
    );
    for (const ev of events) await db.removeEvent(ev.id);
  }
  await prisma.user.deleteMany({
    where: {
      email: testUsers[0].email,
    },
  });
  await prisma.user.deleteMany({
    where: {
      email: testUsers[1].email,
    },
  });
  await prisma.user.deleteMany({
    where: {
      email: testUsers[2].email,
    },
  });
  const jobs: Promise<any>[] = [];
  testUsers.forEach((user) =>
    jobs.push(
      db
        .insertUser(user.email, user.name, user.password)
        .then((returnedUser) => (user.id = returnedUser?.id ?? user.id)),
    ),
  );

  await Promise.all(jobs);

  return await testEvents.forEach((event) =>
    jobs.push(
      createEventStandalone(event, testEventsOwner.id).then(
        (createdEvent) => (event.id = createdEvent?.id ?? event.id),
      ),
    ),
  );
}, 20000);

afterAll(async () => {
  const eventJobs: Promise<any>[] = [];
  testEvents.forEach((event) => {
    eventJobs.push(db.removeEvent(event.id));
  });
  await Promise.all(eventJobs);

  const jobs: Promise<any>[] = [];

  testUsers.forEach((user) => jobs.push(db.removeUser(user.id)));
  return await Promise.all(jobs);
}, 20000);

describe("Database users", () => {
  it("inserts, fetches and deletes a user", async () => {
    const testUser: User = {
      id: "",
      name: "TestUser123",
      email: "testuser@test.com",
      password: "sef098wq3912",
    };
    await db.insertUser(testUser.email, testUser.name, testUser.password);
    const insertedUser = await db.getUserFromEmail(testUser.email);
    expect(insertedUser).toBeTruthy();

    testUser.id = (insertedUser as User).id;
    expect(insertedUser).toEqual(testUser);

    await db.removeUser(testUser.id);
    const removedUser = await db.getUserFromEmail(testUser.email);
    expect(removedUser).toBeFalsy();
  }, 20000);

  it("fetches users based on exact name", async () => {
    const exactUsers = await db.getUsersFromNameOrEmail(
      testUsers[0].name,
      5,
      true,
    );
    for (const user of exactUsers) {
      expect(user.name).toBe(testUsers[0].name);
    }
  }, 20000);

  it("fetches results from multiple search strings", async () => {
    const result = await db.getUsersFromNameOrEmail(
      testUsers.map((user) => user.name),
      5,
      true,
    );
    for (const user of result) {
      user.id = "";
    }
    expect(result).toMatchSnapshot();
    const result2 = await db.getUsersFromNameOrEmail(
      testUsers.map((user) => user.name.split("_")[1]),
      5,
      false,
    );
    for (const user of result2) {
      user.id = "";
    }
    expect(result2).toMatchSnapshot();
  }, 20000);

  it("fetches users based on exact email", async () => {
    const exactUsers = await db.getUsersFromNameOrEmail(
      testUsers[0].email,
      5,
      true,
    );
    expect(exactUsers[0].id).toBe(testUsers[0].id);
  }, 20000);

  it("ignores non-exact email searches", async () => {
    const nonExactUsers = await db.getUsersFromNameOrEmail(
      "@test.example.org",
      5,
      false,
    );
    expect(nonExactUsers.map((user) => user.name).sort()).toEqual([]);
  }, 20000);
});

describe("Database events", () => {
  it("fetches an event with exact name", async () => {
    const eventName = testEvents[0].name;
    const res = await db.getEventsFromName(
      eventName,
      1,
      true,
      testEventsOwner.id,
    );
    expect(res.length).toBe(1);
    expect(res[0].name).toBe(eventName);
  }, 20000);

  it("fetches multiple events with prefix", async () => {
    const eventName = "CS9999 Introduction to Testing - Tutorial";
    const res = await db.getEventsFromName(
      eventName,
      5,
      false,
      testEventsOwner.id,
    );
    for (const event of res) {
      purgeId(event);
    }
    expect(res).toMatchSnapshot();
  }, 20000);

  it("fetches multiple events from id", async () => {
    const eventName = "CS9999 Introduction to Testing - Tutorial";
    const res = await db.getEventMany(testEvents.map((event) => event.id));
    for (const event of res) {
      purgeId(event);
    }
    expect(res).toMatchSnapshot();
  }, 20000);

  it("filters out events with no permission", async () => {
    const eventName = "CS9999 Introduction to Testing - Tutorial";
    const res = await db.getEventsFromName(eventName, 5, false, "");
    expect(res).toEqual([]);
  }, 20000);

  it("fetches multiple events with case insensitive words", async () => {
    const eventName = "cs9999 sem 1";
    const res = await db.getEventsFromName(
      eventName,
      5,
      false,
      testEventsOwner.id,
    );
    for (const event of res) {
      purgeId(event);
    }
    expect(res).toMatchSnapshot();
  }, 20000);

  it("fetches owned events", async () => {
    const res = await db.getOwnedEvents(testUsers[0].id, 20, 1);
    for (const event of res) {
      event.id = "";
      event.ownerId = "";
    }
    expect(res).toMatchSnapshot();
  });
});

describe("Database attendances", () => {
  it("inserts, fetches and deletes an attendance", async () => {
    const event0 = await db.getEventFull(testEvents[0].id);
    const event1 = await db.getEventFull(testEvents[1].id);
    await db.insertAttendance(
      testUsers[0].id,
      event0.id,
      event0.schedules[0]?.id,
    );
    await db.insertAttendance(
      testUsers[1].id,
      event1.id,
      event0.schedules[0]?.id,
    );
    const attendances1 = await db.getAttendancesByUserId(testUsers[0].id);
    for (const x of attendances1) purgeIdAttendance(x);
    expect(attendances1).toMatchSnapshot();
    const attendances2 = await db.getAttendancesByUserIdMany([
      testUsers[0].id,
      testUsers[1].id,
    ]);
    expect(attendances2).toContainEqual({
      attendeeId: testUsers[0].id,
      eventId: event0.id,
      scheduleId: event0.schedules[0]?.id,
    });
    expect(attendances2).toContainEqual({
      attendeeId: testUsers[1].id,
      eventId: event1.id,
      scheduleId: event0.schedules[0]?.id,
    });
    await db.removeAttendance(testUsers[0].id, event0.id);
    await db.removeAttendance(testUsers[1].id, event1.id);
    const attendances3 = await db.getAttendancesByUserIdMany([
      testUsers[0].id,
      testUsers[1].id,
    ]);
    expect(attendances3).toEqual([]);
  }, 20000);

  it("updates attendances", async () => {
    const event3 = await db.getEventFull(testEvents[3].id);
    await db.insertAttendance(
      testUsers[0].id,
      event3.id,
      event3.schedules[0]?.id,
    );
    const attendances1 = await db.getAttendancesByUserId(testUsers[0].id);
    for (const x of attendances1)
      expect(x.scheduleId).toEqual(event3.schedules[0]?.id);
    await db.updateAttendance(
      testUsers[0].id,
      event3.id,
      event3.schedules[1]?.id,
    );
    const attendances2 = await db.getAttendancesByUserId(testUsers[0].id);
    for (const x of attendances2)
      expect(x.scheduleId).toEqual(event3.schedules[1]?.id);
    await db.removeAttendance(testUsers[0].id, event3.id);
  });
});

describe("Database follows", () => {
  it("inserts, fetches and deletes a follow", async () => {
    await db.insertFollow(testUsers[0].id, testUsers[1].id);
    const res1a = await db.getFollowsByFollowerId(testUsers[0].id);
    const res1b = await db.hasRelationship(testUsers[0].id, testUsers[1].id);
    for (const follow of res1a) follow.id = "";
    expect(res1a).toMatchSnapshot();
    expect(res1b).toEqual(false);
    await db.updateFollowStatus(testUsers[0].id, testUsers[1].id, "active");
    const res2a = await db.getFollowsByFollowedId(testUsers[1].id);
    const res2b = await db.hasRelationship(testUsers[0].id, testUsers[1].id);
    for (const follow of res2a) follow.id = "";
    expect(res2a).toMatchSnapshot();
    expect(res2b).toEqual(true);
    await db.removeFollow(testUsers[0].id, testUsers[1].id);
    const res3a = await db.getFollowsByFollowerId(testUsers[0].id);
    const res3b = await db.hasRelationship(testUsers[0].id, testUsers[1].id);
    for (const follow of res3a) follow.id = "";
    expect(res3a).toMatchSnapshot();
    expect(res3b).toEqual(false);
  }, 20000);
});

describe("Database notes", () => {
  it("inserts, fetches and deletes a note", async () => {
    const note = await db.insertNote(
      testUsers[0].id,
      "Test note",
      new Date(1704085200000),
      new Date(1704081600000),
    );
    await db.insertRecipients(note.id, [testUsers[0].id, testUsers[1].id]);
    const res1 = await db.getRecvNotes(testUsers[0].id);
    for (const note of res1 ?? []) {
      note.note.id = "";
      note.note.sender.id = "";
    }
    const res2 = await db.getRecvNotes(testUsers[1].id);
    for (const note of res2 ?? []) {
      note.note.id = "";
      note.note.sender.id = "";
    }
    expect(res1).toMatchSnapshot();
    expect(res2).toMatchSnapshot();
    await db.removeRecv(testUsers[0].id, note.id);
    const res3 = await db.getRecvNotes(testUsers[0].id);
    for (const note of res3 ?? []) {
      note.note.id = "";
      note.note.sender.id = "";
    }
    const res4 = await db.getRecvNotes(testUsers[1].id);
    for (const note of res4 ?? []) {
      note.note.id = "";
      note.note.sender.id = "";
    }
    expect(res3).toMatchSnapshot();
    expect(res4).toMatchSnapshot();
    await db.removeRecv(testUsers[1].id, note.id);
  }, 20000);
});
