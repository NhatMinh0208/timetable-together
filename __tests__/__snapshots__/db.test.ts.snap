// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Database events fetches multiple events from id 1`] = `
[
  {
    "description": "A test case for fetching multiple events with a prefix",
    "id": "",
    "name": "CS9999 Introduction to Testing - Tutorial (Sem 1)",
    "owner": {
      "id": "",
      "name": "user_f8f8fsdf0",
    },
    "private": true,
    "schedules": [
      {
        "id": "",
        "name": "Test Schedule #1",
        "sessions": [
          {
            "endDate": 2024-11-13T00:00:00.000Z,
            "endTime": 2024-01-01T10:00:00.000Z,
            "id": "",
            "interval": 7,
            "place": "Test Location",
            "scheduleId": "",
            "startDate": 2024-07-13T00:00:00.000Z,
            "startTime": 2024-01-01T01:00:00.000Z,
            "timeZone": "",
          },
        ],
      },
    ],
  },
  {
    "description": "A test case for fetching an event with an exact name",
    "id": "",
    "name": "CS9999 Introduction to Testing - Lecture (Sem 1)",
    "owner": {
      "id": "",
      "name": "user_f8f8fsdf0",
    },
    "private": false,
    "schedules": [
      {
        "id": "",
        "name": "Test Schedule #1",
        "sessions": [
          {
            "endDate": 2024-11-13T00:00:00.000Z,
            "endTime": 2024-01-01T10:00:00.000Z,
            "id": "",
            "interval": 7,
            "place": "Test Location",
            "scheduleId": "",
            "startDate": 2024-07-13T00:00:00.000Z,
            "startTime": 2024-01-01T01:00:00.000Z,
            "timeZone": "",
          },
        ],
      },
    ],
  },
  {
    "description": "A test case for fetching multiple events with a prefix",
    "id": "",
    "name": "CS9999 Introduction to Testing - Tutorial (Sem 2)",
    "owner": {
      "id": "",
      "name": "user_f8f8fsdf0",
    },
    "private": true,
    "schedules": [
      {
        "id": "",
        "name": "Test Schedule #1",
        "sessions": [
          {
            "endDate": 2024-11-13T00:00:00.000Z,
            "endTime": 2024-01-01T10:00:00.000Z,
            "id": "",
            "interval": 7,
            "place": "Test Location",
            "scheduleId": "",
            "startDate": 2024-07-13T00:00:00.000Z,
            "startTime": 2024-01-01T01:00:00.000Z,
            "timeZone": "",
          },
        ],
      },
    ],
  },
]
`;

exports[`Database events fetches multiple events with case insensitive words 1`] = `
[
  {
    "id": "",
    "name": "CS9999 Introduction to Testing - Lecture (Sem 1)",
    "owner": {
      "name": "user_f8f8fsdf0",
    },
  },
  {
    "id": "",
    "name": "CS9999 Introduction to Testing - Tutorial (Sem 1)",
    "owner": {
      "name": "user_f8f8fsdf0",
    },
  },
]
`;

exports[`Database events fetches multiple events with prefix 1`] = `
[
  {
    "id": "",
    "name": "CS9999 Introduction to Testing - Tutorial (Sem 1)",
    "owner": {
      "name": "user_f8f8fsdf0",
    },
  },
  {
    "id": "",
    "name": "CS9999 Introduction to Testing - Tutorial (Sem 2)",
    "owner": {
      "name": "user_f8f8fsdf0",
    },
  },
]
`;

exports[`Database follows inserts, fetches and deletes a follow 1`] = `
[
  {
    "email": "yonghu22@test.example.org",
    "id": "",
    "name": "用户_bcvnft908nf",
    "status": "pending",
  },
]
`;

exports[`Database follows inserts, fetches and deletes a follow 2`] = `
[
  {
    "email": "user22@test.example.com",
    "id": "",
    "name": "user_f8f8fsdf0",
    "status": "active",
  },
]
`;

exports[`Database follows inserts, fetches and deletes a follow 3`] = `[]`;

exports[`Database notes inserts, fetches and deletes a note 1`] = `
[
  {
    "note": {
      "content": "Test note",
      "id": "",
      "position": 2024-01-01T05:00:00.000Z,
      "sender": {
        "id": "",
        "name": "user_f8f8fsdf0",
      },
      "timeSent": 2024-01-01T04:00:00.000Z,
    },
    "read": true,
  },
]
`;

exports[`Database notes inserts, fetches and deletes a note 2`] = `
[
  {
    "note": {
      "content": "Test note",
      "id": "",
      "position": 2024-01-01T05:00:00.000Z,
      "sender": {
        "id": "",
        "name": "user_f8f8fsdf0",
      },
      "timeSent": 2024-01-01T04:00:00.000Z,
    },
    "read": false,
  },
]
`;

exports[`Database notes inserts, fetches and deletes a note 3`] = `[]`;

exports[`Database notes inserts, fetches and deletes a note 4`] = `
[
  {
    "note": {
      "content": "Test note",
      "id": "",
      "position": 2024-01-01T05:00:00.000Z,
      "sender": {
        "id": "",
        "name": "user_f8f8fsdf0",
      },
      "timeSent": 2024-01-01T04:00:00.000Z,
    },
    "read": false,
  },
]
`;

exports[`Database users fetches results from multiple search strings 1`] = `
[
  {
    "id": "",
    "name": "user_f8f8fsdf0",
  },
  {
    "id": "",
    "name": "người dùng_hjmghj09m",
  },
  {
    "id": "",
    "name": "用户_bcvnft908nf",
  },
]
`;

exports[`Database users fetches results from multiple search strings 2`] = `
[
  {
    "id": "",
    "name": "user_f8f8fsdf0",
  },
  {
    "id": "",
    "name": "người dùng_hjmghj09m",
  },
  {
    "id": "",
    "name": "用户_bcvnft908nf",
  },
]
`;
